main.cpp

```diff
@@ -13,54 +13,37 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-#include "mbed.h"
 #include <stdio.h>
-#include <algorithm>
-
-// Block devices
-#if COMPONENT_QSPIF
-#include "QSPIFBlockDevice.h"
-QSPIFBlockDevice bd;
-#endif
-
-#if COMPONENT_OSPIF
-#include "OSPIFBlockDevice.h"
-OSPIFBlockDevice bd;
-#endif
-
-#if COMPONENT_SPIF
-#include "SPIFBlockDevice.h"
-// Physical block device, can be any device that supports the BlockDevice API
-SPIFBlockDevice bd;
-#endif
-
-#if COMPONENT_DATAFLASH
-#include "DataFlashBlockDevice.h"
-DataFlashBlockDevice bd;
-#endif
-
-#if COMPONENT_SD
-#include "SDBlockDevice.h"
-// Physical block device, can be any device that supports the BlockDevice API
-SDBlockDevice bd;
-#endif
-
-#include "HeapBlockDevice.h"
+#include "blockdevice/BlockDevice.h"
+#include "blockdevice/HeapBlockDevice.h"
 
 // Entry point for the example
 int main() {
     printf("--- Mbed OS block device example ---\n");
 
+    BlockDevice *bd = BlockDevice::get_default_instance();
+    if(!bd) {
+        printf("No hardware BlockDevice found, using heap\n");
+        // Emulate two 32-byte blocks
+        HeapBlockDevice heap_bd(64, 32);
+        bd = &heap_bd;
+    }
+
     // Initialize the block device
-    printf("bd.init()\n");
-    int err = bd.init();
-    printf("bd.init -> %d\n", err);
+    printf("bd->init()\n");
+    int err = bd->init();
+    printf("bd->init -> %d\n", err);
```

Comment:
We are changing a lot of print statements. Do we also need to update the expected log file?

https://github.com/ARMmbed/mbed-os-example-blockdevice/blob/master/tests/blockdevice.log